src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+                    ApplyOldState(oldState, newItems);
+                }
+            }
+
+            static void AddPreviousState(
+                ITextSnapshot newSnapshot,
+                ImmutableArray<DocumentSymbolDataViewModel> oldItems,
+                PooledDictionary<Span, (bool isExpanded, bool isSelected)> oldState)
+            {
+                foreach (var item in oldItems)
+                {
+                    // EdgeInclusive so that if we type on the end of an existing item it maps forward to the new full span.
+                    var mapped = item.Data.SelectionRangeSpan.TranslateTo(newSnapshot, SpanTrackingMode.EdgeInclusive);
+                    oldState[mapped.Span] = (item.IsExpanded, item.IsSelected);
+
+                    AddPreviousState(newSnapshot, item.Children, oldState);
+                }
+            }
+
+            static void ApplyOldState(
+                PooledDictionary<Span, (bool isExpanded, bool isSelected)> oldState,
+                ImmutableArray<DocumentSymbolDataViewModel> newItems)
+            {
+                foreach (var item in newItems)
+                {
+                    if (oldState.TryGetValue(item.Data.SelectionRangeSpan.Span, out var oldValues))
+                    {
+                        item.IsExpanded = oldValues.isExpanded;
+                        item.IsSelected = oldValues.isSelected;
+                    }
+
+                    ApplyOldState(oldState, item.Children);
+                }
+            }
+        }
+
+        private async Task<(ImmutableArray<DocumentSymbolData> documentSymbolData, ITextSnapshot newTextSnapshot)> ComputeDocumentSymbolDataAsync(CancellationToken cancellationToken)
+        {
+            var filePath = _textBuffer.GetRelatedDocuments().FirstOrDefault(static d => d.FilePath is not null)?.FilePath;
+            if (filePath != null)
+            {
+                // Obtain the LSP response and text snapshot used.
+                var response = await DocumentSymbolsRequestAsync(
+                    _textBuffer, _languageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+                if (response != null)
+                {
+                    var newTextSnapshot = response.Value.snapshot;
+                    var documentSymbolData = CreateDocumentSymbolData(response.Value.response, newTextSnapshot);
+                    return (documentSymbolData, newTextSnapshot);
+                }
+            }
+
+            return (ImmutableArray<DocumentSymbolData>.Empty, _textBuffer.CurrentSnapshot);
+        }
+
+        public void ExpandAndSelectItemAtCaretPosition()
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+            _selectionQueue.AddWork(cancelExistingWork: true);
+        }
+
+        private async ValueTask UpdateSelectionAsync(CancellationToken cancellationToken)
+        {
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
Please verify that there are no delays when scrolling via the mouse cursor. It appears that on every keystroke we are going to jump to the UI thread and attempt to update the selection logic. It does appear that this is necessary as we are trying to avoid re-entrancy but I am worried that the performance of this design will not be acceptable enought to ship.